Add configurable support

PierrickBrun opened the pull request:
Following the example of the bundle product, empty methods are created in order to be inherited.
That way the "flat" default system stays the same and the additional module will allow to import the configurable products in the Odoo format


- [x] import product attribute

- [x] import product attributes values

- [x] import product attribute lines

- [x] change product template

- [x] remove useless product templates
- [x] import product attribute prices
- [x] find a way to hide the principal product.product (linked to the magento configurable product) without removing the magento.product.product associated
example on the first line : ![configurable product.product](https://screenshots.firefoxusercontent.com/images/81f8df79-e129-4dbf-acf5-2ba9822d6074.png)
- [x] unit tests

PierrickBrun committed:
[WIP] add configurable inheritage

- import product attribute OK

- import product attributes values OK

- import product attribute lines OK

- change product template OK

- remove useless product templates WIP

TODO :
- too many unlink product.template jobs
- find a way to hide the principal product.product (linked to the
magento configurable product) without removing the
magento.product.product associated
- unit tests

PierrickBrun committed:
[WIP] add configurable inheritage

- import product attribute OK

- import product attributes values OK

- import product attribute lines OK

- change product template OK

- remove useless product templates WIP

TODO :
- too many unlink product.template jobs
- find a way to hide the principal product.product (linked to the
magento configurable product) without removing the
magento.product.product associated
- unit tests

bealdav commented:
Hi @gurneyalex 
Pierrick is a new talentuous kid on the akretion block (but completely new in odoo) . Could you add it in the akretion team/cla ? Thanks.

bealdav commented:
@guewen 
A said Pierrick'll contribute on Magento for configurable products.
He have to learn a lot here but he dive in your code for its first job in Odoo. It seems he like it
We now have needs to decide what is the way to deal with configurable products.

Our plan is (please fix us):
- import configurable as simple product : almost DONE
- convert this simple product to a full set of product variants (with attributes)
- inactive useless variant

The main question here is: step 2 and 3 should be synchro or deferred with a new job ?

cc @sebastienbeau

guewen commented:
> The main question here is: step 2 and 3 should be synchro or deferred with a new job ?

As step 1 alone doesn't make sense without 2 and 3, it seems to me that it can be synchronous.

guewen commented on connector_magento_configurable/models/inherit.py:
```diff
@@ -0,0 +1,70 @@
+            Allows to delay an unlink job.
+            This is used to delete the orphaned product templates
+        """
+        self.unlink()
```
A job should be idempotent: if run twice, should not fail. Here, you should just check if the product doesn't exist, ignore the unlink.

PierrickBrun replied:
Obviously, yes

guewen commented on connector_magento_configurable/models/product_attribute_line/importer.py:
```diff
@@ -0,0 +1,133 @@
+            )
+        self.env['product.template'].search([
+            ('product_variant_ids', '=', False)
+        ]).with_delay(priority=101, eta=60).delayable_unlink()
```
Is there a reason for giving an eta in 60 seconds?

PierrickBrun replied:
The goal is to avoid trying to unlink a template still having product.product linked to it. Second thought, not sure if this is necessary but didn't think of another way to avoid the problem.

guewen reviewed (commented):
A few details spotted on quick eyeballing

PierrickBrun committed:
[IMP] product attribute prices

- switch to synchronous
- problem : in magento, products created before a change in attribute prices don't change their price. In odoo they do -> can result in prices mismatch

PierrickBrun committed:
[IMP] product attribute prices

- switch to synchronous
- problem : in magento, products created before a change in attribute prices don't change their price. In odoo they do -> can result in prices mismatch

PierrickBrun commented:
Will start testing tomorrow, I think the functional perimeter is OK.

@guewen Do you have an idea how to hide the product.product without attributes ? (screenshot in the first post). If i disable it, the connector will stop updating the template.

Also, what is your opinion on the problem in product attribute prices ?
In magento, products created before a change in attribute prices don't change their price. In odoo they do -> can result in prices mismatch
Is an option to (de)activate it necessary or do we rely on sale order to handle this possible mismatch ?

guewen commented:
> @guewen Do you have an idea how to hide the product.product without attributes ? (screenshot in the first post). If i disable it, the connector will stop updating the template.

What is needed to update on the template? I'm really far to have the whole context, so I can be totally wrong, but I see these options:
* use another 'active' field on the binding which overrides the template's one
* consider even the inactive templates on the code that "update the template" (but I don't what you speak about)



> Also, what is your opinion on the problem in product attribute prices ?
> In magento, products created before a change in attribute prices don't change their price. In odoo they do -> can result in prices mismatch
> Is an option to (de)activate it necessary or do we rely on sale order to handle this possible mismatch ?

I don't know what this problem is... In any way there must be a check in the sale order (it exists I think).

PierrickBrun committed:
[FIX] improvements & tests
 - security/ir.model.access.csv fixed
 - better handling of the from_date filter
 - allows easy override of the sku field
 - add test

PierrickBrun committed:
[FIX] improvements & tests
 - security/ir.model.access.csv fixed
 - better handling of the from_date filter
 - allows easy override of the sku field
 - add test

guewen reviewed (approved):
Thanks!

guewen reviewed (changes requested):
Oops, I reviewed only the last commit. What's the status, still WIP?

guewen commented on connector_magento_configurable/models/configurable.py:
```diff
@@ -0,0 +1,66 @@
+    def run(self, filters=None):
+        """ Run the synchronization """
+        from_date = filters.pop('from_date', None)
+        # to_date = filters.pop('to_date', None)
```
is it something which need to be done?

guewen commented on connector_magento_configurable/models/configurable.py:
```diff
@@ -0,0 +1,66 @@
+        if from_date:
+            criterias.append(('write_date', '>', from_date.strftime(
+                tools.misc.DEFAULT_SERVER_DATETIME_FORMAT)))
+        internal_ids = self.env['magento.product.product'].search(criterias)
```
preferable to give another name than something_ids when it holds records and not ids :)

guewen commented on connector_magento_configurable/models/configurable.py:
```diff
@@ -0,0 +1,66 @@
+        criterias = [('product_type', '=', 'configurable')]
+        if from_date:
+            criterias.append(('write_date', '>', from_date.strftime(
+                tools.misc.DEFAULT_SERVER_DATETIME_FORMAT)))
```
Instead of `strftime(tools.misc.DEFAULT_SERVER_DATETIME_FORMAT)` you should use `fields.DateTime.to_string()`

guewen commented on connector_magento_configurable/models/product_attribute_line/importer.py:
```diff
@@ -0,0 +1,139 @@
+        self._import_record(line)
+        self.env['product.template'].search([
+            ('product_variant_ids', '=', False)
+        ]).unlink()
```
It seems a bit risky to do that. How do you know you are deleting only what you should? There is for instances addons which changes the default behavior, variants are not created automatically and you have to create them with a button. Would be catastrophic here.

guewen reviewed (commented):
Looks good overall. Do you plan to add tests?

bealdav commented:
Yes @PierrickBrun planned to do some tests

PierrickBrun commented:
Thanks for the feedback, yes there will be tests

PierrickBrun committed:
[IMP] Add flag on product.product created from a configurable parent (for easier filtering)

PierrickBrun committed:
[IMP] Add flag on product.product created from a configurable parent (for easier filtering)

PierrickBrun committed:
[IMP] adds possibility to search for inactive product.product as configurable parent

PierrickBrun committed:
[IMP] adds possibility to search for inactive product.product as configurable parent

PierrickBrun committed:
add possibility to store attribute name/code in different fields easily

PierrickBrun committed:
add possibility to store attribute name/code in different fields easily

PierrickBrun committed:
fix use another field for attribute code

PierrickBrun committed:
fix use another field for attribute code

PierrickBrun committed:
fix use another name field

PierrickBrun committed:
fix use another name field

sebastienbeau commented on connector_magento_configurable/models/product_attribute_line/importer.py:
```diff
@@ -0,0 +1,146 @@
+        magento_product.write(
+            {'product_tmpl_id': tmpl_id,
+             'attribute_value_ids': value_ids})
+        magento_product.odoo_id.write(
```
Not needed writing on magento_product will write on magento_product.odoo_id

PierrickBrun committed:
[FIX] better handling of templates unlink

PierrickBrun committed:
[FIX] better handling of templates unlink

PierrickBrun committed:
fix flake8

PierrickBrun committed:
Merge branch '10-configurable' of https://github.com/akretion/connector-magento into 10-configurable

PierrickBrun commented:
Hi @guewen, I think everything is OK now, can you confirm ?

PierrickBrun committed:
[FIX] remove useless parameter

PierrickBrun committed:
[FIX] only unlink selected templates if they are not referenced

PierrickBrun commented:
relaunching travis

PierrickBrun committed:
[FIX] allowed the creation of several product.attribute.line for a unique template/attribute couple

PierrickBrun committed:
[WIP] add magento.product.template

PierrickBrun committed:
[IMP] big refactoring : fixed conflicts with ERP and added support of magento.product.template

PierrickBrun committed:
[FIX] product.attribute.lines not created

PierrickBrun committed:
[IMP] add magento fields on product_template view

PierrickBrun committed:
[FIX] use magento.product.template in tests

PierrickBrun committed:
[FIX] tests

PierrickBrun committed:
[FIX] small fixes

PierrickBrun committed:
fix conflict

sebastienbeau commented on connector_magento_configurable/models/product_product/importer.py:
```diff
@@ -0,0 +1,24 @@
+    def _update(self, binding, data):
+        """ Update an OpenERP record """
+        if binding._name == 'magento.product.product':
+            if binding.product_tmpl_id.magento_bind_ids:
```
if binding.product_variant_count > 1: is better

sebastienbeau committed:
[FIX] setting the variable "variant_managed_by_magento" should be done only at the creation

sebastienbeau committed:
[IMP] add hack to not import description
